src/dsp_tools/commands/xmlupload/xmlupload.py

```diff
@@ -123,6 +129,16 @@ def xmlupload(
     return cleanup_upload(upload_state)
 
 
+def _validate_iiif_uris(root: etree._Element) -> None:
+    uris = [node.text for node in root.iter(tag="iiif-uri")]
+    all_uris = cast(list[str], uris)
+    problems = [res for x in all_uris if (res := IIIFUriValidator(x, is_iiif_uri(x)).validate())]
```

Comment:
This line surfaces multiple problems with this validator abstraction:

1. IMO a client should be a singleton, so you should not have to create a client for each URL you validate, but a single one should be able to validate all of them.
2. If the validator is responsible for validating, then it shouldn't be on the caller side, but in the `validate()` method, where you do the regex check.
3. If there is a validator for each URL you validate, then there is formally no difference between what you pass in to the constructor and what you pass into the `validate()` function. Hence you could isomorphically rewrite this to `IIIFUriValidator().validate(x, is_iiif_uri(x))`. And this proves that like this, there is no advantage of having this be a class - it could just be a top-level function `validate_iiif_uri(x)`.

Your code works equally well to what I suggest, but I think the architecture would be more consistent if the concept of a "client" would always mean the same.

Reply:
This argument makes sense, I will change it as suggested.